zproject/backends.py

```diff
@@ -3517,31 +3514,158 @@ def get_key_and_secret(self) -> tuple[str, str]:
+        request_data = self.strategy.request_data().dict()
+        data_to_relay = {key: request_data[key] for key in params_to_relay if key in request_data}
+        data_to_relay["idp"] = idp_name
+
+        # Associate the important data, especially the IdP which is being used, with the state token
+        # to ensure we have the right set of values for this auth session.
+        # This is more robust than relying on storing variables in request.session.
+        self.put_data_in_redis(state, data_to_relay)
+
+        return state
+
     @classmethod
-    def check_config(cls) -> bool:
-        if len(settings.SOCIAL_AUTH_OIDC_ENABLED_IDPS.keys()) != 1:
-            # Only one IdP supported for now.
-            return False
+    def put_data_in_redis(cls, state: str, data_to_relay: dict[str, Any]) -> str:
+        return put_dict_in_redis(
+            redis_client,
+            "oidc_state_{token}",
+            data_to_store=data_to_relay,
+            expiration_seconds=cls.REDIS_EXPIRATION_SECONDS,
+            token=state,
+        )
 
+    @classmethod
+    def get_data_from_redis(cls, state: str) -> dict[str, Any] | None:
+        data = get_dict_from_redis(
+            redis_client, key_format="oidc_state_{token}", key=f"oidc_state_{state}"
+        )
+
+        return data
+
+    @override
+    def auth_complete(self, *args: Any, **kwargs: Any) -> HttpResponse | None:
+        """
+        Forked from the superclass in order to support multiple IdPs being enabled on the server.
+        Needs to safely determine:
+        1) Which IdP is handling this authentication attempt
+        2) Which subdomain we're logging into
+        3) Validate that the IdP is allowed for authentication users for this subdomain.
+        """
+        self.process_error(self.data)
+        state = self.validate_state()
+        assert state
+
+        relayed_params = self.get_data_from_redis(state)
+        assert relayed_params is not None
+        for param in self.standard_relay_params:
+            relayed_value = relayed_params.get(param)
+            session_value = self.strategy.session_get(param)
+            # A hardening measure. These params are automatically saved in the user's Zulip session
+            # by python-social-auth when auth is initiated. More robustly, we associate these values
+            # with the "state" token generated for this OIDC authentication attempt and store that in redis.
+            # It's this latter storage that we want to rely on for fetching these values here in auth_complete,
+            # after the user has been redirected back to us from the IdP.
+            # However, if all is right, then the values in the session should match what we retrieve from redis.
+            assert relayed_value == session_value
+
+        idp_name = relayed_params["idp"]
+        assert idp_name
+        subdomain = relayed_params["subdomain"]
+        assert subdomain is not None
```

Comment:
Where are we validating that the request is coming on for the relayed subdomain? Is it that we're going to redirect to login to that `subdomain` once auth is successful? Just double-checking my understanding.

Reply:
yes, the user comes to `auth.zulipchat.com/complete/oidc/?<params>` (in multi-org servers, which are supposed to use `SOCIAL_AUTH_SUBDOMAIN`) and then we redirect to the intended `subdomain` at the end of auth, based on the `subdomain` value we stored at the beginning of auth. This is how it works for all python-social-auth backends for us.